openmpi.spec: updates for v5.0.0

jsquyres opened the pull request:
- Default all 3rd-party packages to be "external".  It is strognly
  recomended that the RPM only contain Open MPI, not the embedded
  3rd-party packages.  This behavior can be overridden via the
  configure_options rpmbuild macros.
- Copy all the Sphinx-generated HTML files to pkgdocdir/html.  These
  are not normally installed, but it's probably worth it for the RPM.
- Remove no-longer-existing INSTALL file.  This file was removed as
  part of the conversion to RST-based documentation.
- Tighten up files listings for the individual RPMs.
- Convert -docs sub-package to exclusively use mandir and pkgdocdir
  (instead of docs.files).  We already excluded mandir because
  rpmbuild may have renamed all the man pages after compressing them
  (so we didn't necessarily know the correct filenames).  We also
  explicitly list pkgdocdir so that it picks up the entire tree of
  files and also removes the entire tree (including directories) when
  the RPM is removed.
- Remove some stale F77 and VT references.
- Fix some minor typos.
- Fix a few "file listed twice" warnings (but leave at least one set
  of these because in some cases it will appear, but in other cases it
  will not -- depending on the prefix).

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

jsquyres commented:
@acgoldma @dancejic You both indicated that your orgs use the specfile.  Please review this PR.

Note a very important change in default behavior in this version of the specfile: it defaults to `--with-hwloc=external --with-libevent=external --with-pmix=external --with-prrte=external`.

If you run into any problems with this update, patches / suggested fixes would be greatly appreciated.  😄

wzamazon commented:
Thank you! We will test the PR.

> --with-hwloc=external --with-libevent=external --with-pmix=external --with-prrte=external.

Is the recommendation  to install pmix and prrte before install openmpi rpm?

bwbarrett commented:
> Thank you! We will test the PR.
> 
> > --with-hwloc=external --with-libevent=external --with-pmix=external --with-prrte=external.
> 
> Is the recommendation to install pmix and prrte before install openmpi rpm?

That is the strong advice, yes.  See https://docs.open-mpi.org/en/v5.0.x/installing-open-mpi/required-support-libraries.html#strong-advice-for-packagers.

shijin-aws commented:
> That is the strong advice, yes. See https://docs.open-mpi.org/en/v5.0.x/installing-open-mpi/required-support-libraries.html#strong-advice-for-packagers.

Does that mean it is recommended to install pmix and prrte via package manager (`yum`, `apt`) from the OS repo?

bwbarrett commented on contrib/dist/linux/openmpi.spec:
```diff
@@ -212,7 +207,14 @@
+# (vs. including other, 3rd party packages in this Open MPI RPM).
+# Meaning: if you set a value to configure_options, you are strongly
+# advised to *also* include all the options listed below.
+%{!?configure_options: %define configure_options --with-libevent=external --with-hwloc=external --with-pmix=external --with-prrte=external}
```
I guess we don't need the requires lines, because the shared library tracing code will do that for us (although I wonder if we should to support a `--enable-static --disable-shared` case, or maybe document said case doesn't work.

It does feel like we could be more user friendly and BuildRequires the devel packages for hwloc and libevent.  But maybe that's a bridge too far?

jsquyres replied:
Fair point.  This gets a little tricky since the user is also able to define `configure_options`, but I think that this script+specfile can't protect the user from themselves in all cases.  However, I'll at least separate the two issues slightly by:

1. Keep `configure_options`, but resume its defaults back to `%{nil}` / `""`
2. Add a boolean `all_external_3rd_party` which will add `--with-libevent=external --with-hwloc=external ...` to the `configure` command line after `%{configure_options}`.

jsquyres replied:
@bwbarrett I pushed a "SQUASHME" commit up here so you can see the changes.  If you're good with it, I'll squash it back down to a single commit.

shijin-aws commented:
@bwbarrett I see. Thank you for your explanation!

acgoldma commented:
I am building the rpms for the prrte and opepmix, but looks like there is an issue with prrte (prte). 
```
# contrib/dist/make_dist_tarball --no-greek --highok --no-git-update
...
=============================================
prrte-3.0.0 archives ready for distribution:
prrte-3.0.0.tar.gz
prrte-3.0.0.tar.bz2
=============================================

#  ./contrib/dist/linux/buildrpm.sh -s ../prrte-3.0.0.tar.bz2
...
--> Created destination specfile: /root/RPMBUILD/SPECS/prte-3.0.0.spec
--> Building the PRTE SRPM
--> rpmbuild --define 'mflags -j4' --define '_source_filedigest_algorithm md5'  --define '_binary_filedigest_algorithm md5' --define '_topdir /root/RPMBUILD' --define 'dist %{nil}' -bs /root/RPMBUILD/SPECS/prte-3.0.0.spec
error: Bad source: /root/RPMBUILD/SOURCES/prte-3.0.0.tar.bz2: No such file or directory
*** FAILURE BUILDING SRPM!
```

Looks to be a simple name issue?

acgoldma commented:
SRPM is building now with this patch to spec file:
```
rpmbuild --define "_name prrte" -bs /root/rpmbuild/SPECS/prte-3.0.0.spec
```
```diff
@@ -355,7 +355,7 @@
 # there that are not meant to be packaged.
 rm -rf $RPM_BUILD_ROOT

-%setup -q -n prte-%{version}
+%setup -q -n %{name}-%{version}

 #############################################################################
 #
```

RPM is failing with unpackaged files.
```
Checking for unpackaged file(s): /usr/lib/rpm/check-files /root/rpmbuild/BUILDROOT/prrte-3.0.0-1.el8.x86_64
error: Installed (but unpackaged) file(s) found:
   /usr/include/prte.h
   /usr/include/prte_version.h


RPM build errors:
    Installed (but unpackaged) file(s) found:
   /usr/include/prte.h
   /usr/include/prte_version.h
```

shijin-aws commented:
I hit this error when generating a ompi  tarball via `make dist`:

```
$  ./autogen.pl
...
$  ./configure --prefix=/opt/openmpi --with-libevent=external --with-hwloc=external --with-prrte=external --with-pmix=external
...
$  make dist

make[3]: Entering directory `/home/ec2-user/ompi/test/spc'
make[3]: Leaving directory `/home/ec2-user/ompi/test/spc'
make[2]: Leaving directory `/home/ec2-user/ompi/test'
 (cd docs && make  top_distdir=../openmpi-gitclone distdir=../openmpi-gitclone/docs \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make[2]: Entering directory `/home/ec2-user/ompi/docs'
make[2]: *** No rule to make target `_build/man/ompi-wrapper-compiler.1', needed by `distdir'.  Stop.
make[2]: Leaving directory `/home/ec2-user/ompi/docs'
make[1]: *** [distdir] Error 1
make[1]: Leaving directory `/home/ec2-user/ompi'
make: *** [dist] Error 2
```
@jsquyres do you have any ideas where the issue is?

jsquyres commented:
@acgoldma If the PMIx and PRRTE packages have problems with their specfiles, you'll need to bring that up in their repos.  

FWIW: I think that PRRTE is really, really weird (and will be very confusing for users) for using `prrte` in some places and `prte` in other places.  I've mentioned this before, but have not won this argument.

jsquyres commented:
@shijin-aws Yes: you need to install Sphinx.  See https://docs.open-mpi.org/en/main/developers/prerequisites.html#sphinx and https://docs.open-mpi.org/en/main/developers/sphinx.html.

acgoldma commented:
Looks like romio341 detects Level-Zero on my system and causes issues. 
```
  CC       src/gavl/mpl_gavl.lo
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c: In function 'MPL_gpu_ipc_handle_create':
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c:123:11: warning: implicit declaration of function 'zeDriverGetMemIpcHandle'; did you mean 'zeMemGetIpcHandle'
? [-Wimplicit-function-declaration]
     ret = zeDriverGetMemIpcHandle(global_ze_driver_handle, ptr, ipc_handle);
           ^~~~~~~~~~~~~~~~~~~~~~~
           zeMemGetIpcHandle
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c: In function 'MPL_gpu_ipc_handle_map':
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c:139:9: warning: implicit declaration of function 'zeDriverOpenMemIpcHandle'; did you mean 'zeMemOpenIpcHandle
'? [-Wimplicit-function-declaration]
         zeDriverOpenMemIpcHandle(global_ze_driver_handle,
         ^~~~~~~~~~~~~~~~~~~~~~~~
         zeMemOpenIpcHandle
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c:140:69: error: 'MPL_gpu_ipc_mem_handle_t' {aka 'struct _ze_ipc_mem_handle_t'} has no member named 'global_dev
_id'
                                  global_ze_devices_handle[ipc_handle.global_dev_id],
                                                                     ^
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c:141:44: error: 'MPL_gpu_ipc_mem_handle_t' {aka 'struct _ze_ipc_mem_handle_t'} has no member named 'handle'
                                  ipc_handle.handle, ZE_IPC_MEMORY_FLAG_NONE, ptr);
                                            ^
../../../../../3rd-party/romio341/mpl/src/gpu/mpl_gpu_ze.c:141:53: error: 'ZE_IPC_MEMORY_FLAG_NONE' undeclared (first use in this function); did you mean 'ZE_IPC_MEMORY
_FLAG_TBD'?
                                  ipc_handle.handle, ZE_IPC_MEMORY_FLAG_NONE, ptr);
                                                     ^~~~~~~~~~~~~~~~~~~~~~~
                                                     ZE_IPC_MEMORY_FLAG_TBD
```

Looks like it is not using the correct API?

I will remove level-zero form my system temporarily, but this should be disabled somehow.

jsquyres commented:
@acgoldma Sounds like that's a general bug in the ROMIO in OMPI master/v5.0.x (i.e., nothing to do with the RPM packaging).  Can you file a separate issue about it and label/milestone it for master+v5.0.x?

shijin-aws commented:
@jsquyres Thanks. That works for me. I could build the ompi rpm and install it successfully. But I had an issue when running mpirun
```
$. /buildrpm.sh -b -R $PWD/ompi_build -r "--define '_prefix /opt/amazon/openmpi'" openmpi-5.1.0a1.tar.bz2
...
$ sudo yum install -y /home/ec2-user/ompi-main/contrib/dist/linux/ompi_build/RPMS/x86_64/openmpi-5.1.0a1-1.amzn2.x86_64.rpm
...
$ /opt/amazon/openmpi/bin/mpirun
The mpirun ("/opt/amazon/openmpi/bin/prterun") cmd failed to exec its actual executable - your application will NOT execute. Error: No such file or directory

$[ec2-user@ip-172-31-62-242 linux]$ which prterun
/usr/local/bin/prterun
```
So I installed prrte earlier in `/usr` . Must I install prrte in the same location with ompi?

acgoldma commented:
> 
> So I installed prrte earlier in `/usr` . Must I install prrte in the same location with ompi?

Same issue, I just added a symlink as a workaround and it seems to work so far, need a bit more testing.

bwbarrett commented:
> $ /opt/amazon/openmpi/bin/mpirun
> The mpirun ("/opt/amazon/openmpi/bin/prterun") cmd failed to exec its actual executable - your application will NOT execute. Error: No such file or directory
> 
> $ which prterun
> /usr/local/bin/prterun
> ```
> 
> So I installed prrte earlier in `/usr` . Must I install prrte in the same location with ompi?

Is PRRTE installed in /usr or /usr/local?  For various reasons, Open MPI requires PRRTE not move after install.  it's weird that it found /usr/local/bin/prterun during RPM build if it was in /usr.  But hard to tell what happened without seeing the config.log of the build.

shijin-aws commented:
@bwbarrett I did not install prrte via rpm, I just `./autogen.pl; ./configure; sudo make install`. The `config.log` shows it has a default prefix as `/usr`

```
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,objc,obj-c++,fortran,
ada,go,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http:/
/bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release
--enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --ena
ble-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash
-style=gnu --enable-plugin --enable-initfini-array --with-isl --enable-libmpx --enable-libsanit
izer --enable-gnu-indirect-function --enable-libcilkrts --enable-libatomic --enable-libquadmath
 --enable-libitm --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
```
And I did not move prrte's location after installation.

acgoldma commented:
I think I found the bug in the configure:
```
config.out:./configure: line 77762: tets: command not found
configure:                  if tets -z "$PRTE_PATH"; then :
config/ompi_setup_prrte.m4:                  AS_IF([tets -z "$PRTE_PATH"],
```

this appears to be a bug here:
https://github.com/open-mpi/ompi/blob/9eec56222a5c98d13790c9ee74877f1562ac27e8/config/ompi_setup_prrte.m4#L287
```diff
-                  AS_IF([tets -z "$PRTE_PATH"],
+                 AS_IF([test -z "$PRTE_PATH"],

awlauria commented:
@acgoldma nice catch - can you please PR?

bwbarrett commented:
> @acgoldma nice catch - can you please PR?

I'll post a PR to fix  it.  There's something else going on beyond just that issue; we should not have gotten into a situation where we're calling the wrong prterun just because of that one typo.

jsquyres commented:
I found another unexpected breakage (i.e., assumptions of OMPI and PRTE and/or PMIx in the same install tree): the Java MPI bindings.  I'll file a separate issue about this.

shijin-aws commented:
> we should not have gotten into a situation where we're calling the wrong prterun just because of that one typo.

Yeah I tried that patch and re-created the ompi tarball and rpm, but the `prte` location issue does not get fix.

bwbarrett commented on contrib/dist/linux/buildrpm.sh:
```diff
@@ -471,7 +468,7 @@ cat <<EOF
 ------------------------------------------------------------------------------
 A copy of the tarball is located in: $rpmtopdir/SOURCES/
-The completed rpms are located in:   $rpmtopdir/RPMS/i<something>86/
+The completed rpms are located in:   $rpmtopdir/RPMS/x86_64/
```
Can we just say RPMS/<architecture>?  Some of us build on ARM :).

shijin-aws replied:
+1

jsquyres replied:
Fixed.

jsquyres commented:
I have rebased on master HEAD (to get the installation of HTML docs), and I believe I have addressed all reviewer comments and also fixed up the contents of the runtime, devel, and docs RPMs.

I think we still have PMIx/PRTE integration issues (e.g., if installation prefixes are different, etc.), but if @acgoldma @shijin-aws @wzamazon @bwbarrett could verify that the all-in-one and/or sub-project RPMs are building like you expect them to / contain what you expect them to, that would be great.  Thanks!

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

jjhursey commented:
bot:ibm:retest

shijin-aws commented:
@jsquyres  I did not hit issue for the Ompi spec file within this PR's range. I hit the RPM build and install issue as @acgoldma reported https://github.com/openpmix/prrte/issues/1332 for PRRTE

shijin-aws commented:
@jsquyres I build/use the all-in-one RPMs (`build_rpms.sh -b`)

acgoldma commented:
> @jsquyres I build/use the all-in-one RPMs (`build_rpms.sh -b`)

Same with me, though we build in 2 steps:  first SRPM (no -b) then use rpmbuild to build single rpm form SRPM.

jsquyres commented:
An extra commit somehow got in this PR ("Remove AMCA param sets") that should not have been in this PR.  Just pushed to remove that one commit and leave just the one RPM spec file update commit.

bwbarrett reviewed (changes requested):
Your commit message looks like it has the remnants of a checkpoint commit squash.
